src/main/java/at/hannibal2/skyhanni/features/garden/farming/GardenCustomKeybinds.kt

```diff
@@ -24,71 +29,121 @@ object GardenCustomKeybinds {
+    private fun update() {
+        with(config) {
+            with(mcSettings) {
+                map = buildMap {
+                    fun add(keyBinding: KeyBinding, property: Property<Int>) {
+                        val value = property.get()
+                        if (value != keyBinding.keyCode) put(keyBinding, value)
+                    }
+                    add(keyBindAttack, attack)
+                    add(keyBindUseItem, useItem)
+                    add(keyBindLeft, left)
+                    add(keyBindRight, right)
+                    add(keyBindForward, forward)
+                    add(keyBindBack, back)
+                    add(keyBindJump, jump)
+                    add(keyBindSneak, sneak)
+                }
             }
-            return false
         }
+        calculateDuplicates()
+        lastDuplicateKeybindsWarnTime = SimpleTimeMark.farPast()
+        KeyBinding.unPressAllKeys()
+    }
 
-        return true
+    private fun calculateDuplicates() {
+        isDuplicate = map.values
+            .filter { it != Keyboard.KEY_NONE }
+            .let { values -> values.size != values.toSet().size }
     }
 
+    private fun isEnabled() = GardenAPI.inGarden() && config.enabled && !(GardenAPI.onBarnPlot && config.excludeBarn)
+
+    private fun isActive(): Boolean =
+        isEnabled() && GardenAPI.toolInHand != null && !isDuplicate && lastWindowOpenTime.passedSince() > 300.milliseconds
+
     @JvmStatic
-    fun isKeyDown(keyBinding: KeyBinding, cir: CallbackInfoReturnable<Boolean>) {
-        if (!isActive()) return
-        val override = map[keyBinding] ?: return
-        val keyCode = override()
-        cir.returnValue = keyCode.isKeyHeld()
+    fun disableAll() {
+        with(config) {
+            attack.set(Keyboard.KEY_NONE)
+            useItem.set(Keyboard.KEY_NONE)
+            left.set(Keyboard.KEY_NONE)
+            right.set(Keyboard.KEY_NONE)
+            forward.set(Keyboard.KEY_NONE)
+            back.set(Keyboard.KEY_NONE)
+            jump.set(Keyboard.KEY_NONE)
+            sneak.set(Keyboard.KEY_NONE)
+            Minecraft.getMinecraft().thePlayer.closeScreen()
+        }
     }
 
     @JvmStatic
-    fun onTick(keyCode: Int, ci: CallbackInfo) {
-        if (!isActive()) return
-        if (keyCode == 0) return
-        val keyBinding = map.entries.firstOrNull { it.value() == keyCode }?.key ?: return
-        ci.cancel()
-        keyBinding as AccessorKeyBinding
-        keyBinding.pressTime_skyhanni++
+    fun defaultAll() {
+        with(config) {
+            attack.set(KeyboardManager.LEFT_MOUSE)
+            useItem.set(KeyboardManager.RIGHT_MOUSE)
+            left.set(Keyboard.KEY_A)
+            right.set(Keyboard.KEY_D)
+            forward.set(Keyboard.KEY_W)
+            back.set(Keyboard.KEY_S)
+            jump.set(Keyboard.KEY_SPACE)
+            sneak.set(Keyboard.KEY_LSHIFT)
+            Minecraft.getMinecraft().thePlayer.closeScreen()
+        }
     }
```

Comment:
I know they didnt have it before but could we have some chat feedback for these, also idk why they exit the config menu so maybe that can be removed but thats up to you. (If the exiting config is removed there doesnt need to be a chat message)

Reply:
done